### ISSUE TITLE
Error writing to a pre-existing file

### ISSUE THREAD

**Issue by clmendes** [LLNL/UnifyFS#488]
As suggested by Adam, I am creating this issue based on email that I had sent to @adammoody  and to @CamStan. The source code and datafile for testing this were "given" to them on Catalyst. Adam suspects he knows what's the root of the problem.

Adam, Cameron:

As I mentioned in today's call, I created a very simple example
that reproduces precisely the error that I was seeing in the
second phase of the HDF5/Unify failing example, where an existing
file is opened and rewritten.

This simple example is independent of HDF; I will "give" it to you
on Catalyst, with the corresponding data file that it uses (also
independent of HDF). This program simply does the following:

         char FILE[128] = "/unifyfs/datafile" ;
         char buf[MAX];                            (MAX=300)
         . . .
         fd = open(FILE,O_RDWR);
         write(fd, buf, 100);
         fprintf(stderr,"WILL SEEK\n");

Execution of this code on Catalyst never reaches the "fprintf", and
generates a failure at the end of the "write". Below I'm showing
what I get on stderr at the point of that "write".

Notice that before running this code, I make sure that "datafile" is
an existing file, with 300 bytes (I will "give" it to you too). Before
running this code, I import the file to Unifyfs via

        srun -n 1 $UNIFYFS_BINDIR/sysio-cp-gotcha $mydir/datafile
/unifyfs/datafile

and then I run the example via

        srun -n 1  $mydir/overwrite-gotcha

where, as shown below, I can see the 'open', the 'write' of
100 bytes, and then the error happens.

Thus, the error, in summary, happens when an existing Unify-hosted
file is opened and then a "write" is done on it. That's why I was
asking if any previous example in Unify does the same.

-Celso

---------------------------------------------------------------

STDERR from execution of the code above:

. . .
2020-04-20T17:21:55 tid=1918 @ __wrap_open() [unifyfs-sysio.c:778]
UNIFYFS_open generated fd 0 for file /unifyfs/datafile
2020-04-20T17:21:55 tid=1918 @ __wrap_write() [unifyfs-sysio.c:1034]
write was called for fd 1024, count 100
2020-04-20T17:21:55 tid=1918 @ unifyfs_intercept_fd() [unifyfs.c:375]
Changing fd from exposed 1024 to internal 0
*** Error in `/g/g12/mendes3/MPI/CELSO/TEST/DIR/overwrite-gotcha':
free(): invalid pointer: 0x0000000001454500 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x816b9)[0x2aaaac4526b9]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_common.so.1(seg_tree_add+0x18b)[0x2aaaaf98579b]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_gotcha.so.1(unifyfs_add_index_entry_to_seg_tree+0xb4)[0x2aaaab2797c4]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_gotcha.so.1(+0x16d17)[0x2aaaab279d17]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_gotcha.so.1(unifyfs_fid_store_fixed_write+0x72)[0x2aaaab27ac02]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_gotcha.so.1(unifyfs_fd_write+0x9c)[0x2aaaab28136c]
/g/g12/mendes3/UnifyFS-Catalyst/UnifyFS/install/lib/libunifyfs_gotcha.so.1(__wrap_write+0x121)[0x2aaaab282021]
/g/g12/mendes3/MPI/CELSO/TEST/DIR/overwrite-gotcha[0x40270d]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x2aaaac3f3545]
/g/g12/mendes3/MPI/CELSO/TEST/DIR/overwrite-gotcha[0x4024a9]
======= Memory map: ========
00400000-0041f000 r-xp 00000000 00:3b 1573622182
/g/g12/mendes3/MPI/CELSO/TEST/DIR/overwrite-gotcha
0061e000-0061f000 r--p 0001e000 00:3b 1573622182
/g/g12/mendes3/MPI/CELSO/TEST/DIR/overwrite-gotcha
. . .

**Comment by adammoody** [comment]
@clmendes , here is a patch that might work around this problem.

https://github.com/LLNL/UnifyFS/pull/487/commits/990e8e69a97a88a79a18773e7574c9a750a27df6

I don't think we can reproduce this problem in the dev branch right now, because the process in the second job does not reattach to the superblock from the process in the first job.

**Comment by adammoody** [comment]
For background info on this thread...

Thanks!  Nice catch!  Can you go ahead and create an issue for this?

I haven't run this, but I have an idea of what's going on.

On writes, we add index entries to a segment tree that is maintained in memory.  We include a pointer to the segment tree in our filemeta data.  I think the open in the second program will end up locating the filemeta data struct that was created by the first run, because that structure is stored in shared memory that the second job can attach to.  However, the pointer to the segment tree is a pointer to data that malloc'd in the first run and is not longer valid memory in the second run.

This is a known problem that will affect both our segment tree and the new client local extents optimization.  Both of these structures call malloc() to allocate memory.  We instead need to change these to either: maintain these structures in shared memory that the second job can load or we need to recreate the trees from data that is stored in shared memory.

**Comment by adammoody** [comment]
As a simple fix, we might be able to just "reset" those pointers when initializing the library for any active file so that they point to empty, newly allocated data structures.  We'll lose performance versus recreating the full data structure that way, but I think it would still be correct and it may be easy to add.  I can look at that later.

**Comment by clmendes** [comment]
@adammoody , @CamStan : I applied this patch (990e8e6) provided by Adam to my (old) Unifyfs version. I repeated execution of the example provided above (program mpi_overwrite.c) rebuilt with this patched version of Unify. My new results contain good and bad news, which I'll report below.

First, the good news: I no longer have the fault at execution, and the example runs through the end,  which means the patch from Adam did its job: the write to the existing file completes without generating the execution fault.

Now, the bad news: the resulting file does not have the correct size. Initially, the pre-existing file had 300 bytes, and the example program consists of just this:

        char FILE[128] = "/unifyfs/datafile" ;
        . . .
        fd = open(FILE,O_RDWR);
        write(fd, buf, 100);
        fprintf(stderr,"WILL SEEK\n");
        lseek (fd, 200, SEEK_SET);
        write(fd, buf, 72);
        close(fd);

The resulting file now has 272 bytes, not 300! In other words, the second "write", which wrote 72 bytes after 'seeking' to position 200, completed fine, and apparently left the "end-of-file" mark there at position 272 when the file was closed. Obviously, the file should still have 300 bytes as it had at the beginning of the execution (that's how the code works without Unify).

I rebuilt and reran the tests of the HDF5 testsuite with this patched Unify, and the programs that reopen/rewrite an HDF file still fail, the same way they failed before (i.e. the resulting file cannot be opened by h5dump).

I strongly suspect this is the same problem that we had originally, i.e. after doing the new write's, somehow the original contents of the file that existed and were not touched are "forgotten" -- that's why the new file in the example above ends up having only 272 bytes. I'm afraid this might be the same mdhim issue that Adam had raised before.

Thus, I'm not sure of how to proceed: we could just close this issue and consider that the patch "fixes" the first write problem (which it really does), but if we do that, then we'd have to open another issue, using the *same* example code (mpi_overwrite.c), to investigate the resulting filesize problem (i.e. possibly an mdhim problem). Or we can leave this issue still open (my preference), because its subject is still valid: the problem is still an effect of writing to a pre-existing file that was reopened.

**Comment by adammoody** [comment]
@clmendes , thanks for the update.  Glad to hear the patch avoids the segfault.  Thanks for trying that out.

I don't know all of the details here, but yes, the wrong file size problem is probably due to the known MDHIM issue.  I suspect the first job that creates the file with the copy writes all 300 bytes starting from offset 0.  That would then produce a single key/value pair in MDHIM representing the byte range:

[0,300)

The second job would then insert two separate key/value pairs for its two writes of [0,100) and [200,272).  That [0,100) key/value pair will end up overwriting the existing [0,300) key/value pair because our MDHIM implementation uses the starting byte offset as the key and we don't bother to look for and rewrite existing key/value pairs.  So we end up with:

[0,100)
[200,272)

instead of what we should have:

[0,100)
[100,200)
[200,272)
[272,300)

**Comment by clmendes** [comment]
@adammoody , yes, I believe it is exactly like you pointed. I import the original datafile into /unifyfs via the "sysio-cp-gotcha" program, which is likely to do the copy of the 300 bytes in a single chunk. After execution of the program, the resulting file ends up having only two valid ranges:

[0,100)
[200,272)

I can confirm this easily from another direction: in my test, the original 300-byte file, created in advance with a text editor, had the constant value 'A' for all the 300 bytes. The program I "gave" you (mpi_overwrite.c) initializes the buffer with 

  for (i=0; i<MAX; i++) buf[i]=(unsigned)i;

After execution under Unify, I check the resulting 272-byte file, and it has the following contents:

[0,100):  i                            (i.e.  0,1,2,3,..)
[100,200):  0 (zero)             (i.e.  0,0,0,...,0)
[200,272):  i-200                (i.e.  0,1,2,3,..)

In other words, besides missing the last 28 bytes (i.e. filesize problem), the resulting file has *different* values for the range [100,200) that was not touched by the two write's in the test program! That's what I meant by saying that the original file's contents are "forgotten" by Unify!

In practice, it is the same that you meant when you said that the range [100,200} was also missing  in the resulting file.

Running this example code outside Unify, I end up with a file that looks like this:

[0,100):       i
[100,200):  A
[200,272):  200-i
[272,300):  A

Thus, the original program (mpi_overwrite.c) and datafile that I "gave" you on Catalyst are still valid to demonstrate this problem as well.

**Comment by clmendes** [comment]
@adammoody , I made additional tests today on Catalyst, with a slightly modified version of the example, which had the following core:

        fd = open(FILE,O_RDWR);
        lseek (fd, 10, SEEK_SET);              <===================
        write(fd, buf, 100);
        fprintf(stderr,"WILL SEEK\n");
        lseek (fd, 200, SEEK_SET);
        write(fd, buf, 72);
        close(fd);

In other words, before the first write, I do a "seek" to position 10. The result from this execution is a file with 300 bytes, with the following contents:

[0:300):  A

Thus, the two "write's" in the program are ignored, and the original contents of the file are preserved. This is probably what happened in that HDF example for 'group-rewrite'. where the original contents (i.e. group name) are preserved.

**Comment by adammoody** [comment]
Thanks again, @clmendes .  I think what happens in this case is that after those two writes, we probably end up with all of the following in MDHIM:

[0, 300)
[10, 110)
[200, 272)

The file size will work out to be 300 here, because we take the max offset across all keys.  The data you actually get back on a read may be undefined, since I think it depends on the order in which the keys come back from MDHIM.  In this case that [0,300) was the one to win out.

**Comment by clmendes** [comment]
@adam, I agree. In my Unify version, I had instrumented function process_read_data()  in file unifyfs-sysio.c, so that the arguments passed to that function are printed via LOGDBG commands. That function contains this:

static int process_read_data(read_req_t* read_reqs, int count, int* done)
. . .
/* get number of read replies in shared memory */
    size_t num = shm_hdr->meta_cnt;

For this latest version of the code (with the seek-to-10 before the write), when I export the resulting data file via 'sysio-cp-gotcha' to obtain the 300-byte file, I see the following in this function:

    count=1,  num=1:   rep-length=300

and the data read (dumped by me to stderr) indeed contains 300 values of  'A'. Thus itlooks like the delegator only sends back one range (0:300)

Meanwhile, in the other version of the code, which produced the 272-byte file, this same function produced the following:

   count=1,  num=2:   rep->length=100,  rep->length=72

Thus, in that case there were two read-replies in shared memory; for both replies, the data had the expected values, i.e. {0,1,2,...,99}  and  {0,1,2,3,...,71}

**Comment by clmendes** [comment]
@adammoody  ,

After getting a fresh clone of 'dev' yesterday (Apr.29,2020), I got slightly different behavior with the two versions of the example mentioned in this issue. To work with this new Unify/dev code on Catalyst, first I had to properly set new env-variables in my job script, namely:

export UNIFYFS_LOG_VERBOSITY=5
export UNIFYFS_LOGIO_SHMEM_SIZE=$(expr 32 \* 1048576)
export UNIFYFS_LOGIO_SPILL_SIZE=$(expr 256 \* 1048576)
export UNIFYFS_LOGIO_SPILL_DIR=$MY_NODE_LOCAL_STORAGE_PATH

With this new setting, I reran the two versions of the example, which reads a pre-existing datafile with 300 bytes. The results were as follows:

a) Code with "lseek":

The core of this version is:

        fd = open(FILE,O_RDWR);
        lseek(fd,10,SEEK_SET);
        write(fd, buf, 100);
        close(fd);

This now *works* fine! The result is a 300-byte file with the expected contents:
* bytes [0-10) come from the original datafile;
* bytes [10-110) come from the 'buf' array in the program;
* bytes [110-300) come from the original datafile;

b) Code without "lseek": 

        fd = open(FILE,O_RDWR);
        write(fd, buf, 100);
        close(fd);

With this version, the result is a 100-byte file, with the contents being the data in the 'buf' array of the code. Thus, the original data between [100-300) are lost!
For this version, function "process_read_data()" (now located in file unifyfs.c) is invoked with count=1, and there is one read-reply (i.e. shm_hdr->meta_cnt=1), with a length of 100 (i.e. rep->length=100).

**Comment by adammoody** [comment]
Hi Celso,
Thanks for the update.  I think we're just lucky that case a) now works.  Under the covers, I think our keys in MDHIM will still be wrong, but we happen to get the right answer because the read path happens to process those keys in the correct order.
-Adam

________________________________________

**Comment by clmendes** [comment]
@adammoody , I agree that we may be lucky, but with the (a) version of the code and the current version of Unify from 'dev' the code works fine. When exporting the file (by reading it) the process_read_data() function gets "count=1" and "num=2", i.e. there are two read replies, with lengths 300 and 100, respectively. 

For case (b), it still failes, as the resulting file has length 100, and there is only one read-reply with length=100, as already described. So, this case (b) can be used in the future when we get a replacement for mdhim.

**Comment by clmendes** [comment]
@adammoody : I modified slightly the example that is used with this issue, to make it easier to be tested. It is not necessary to have a pre-existing file, the program itself can create the initial 300-byte file. Thus, this code becomes self-contained. The core of the code then becomes:

        fd = open(FILE,O_CREAT|O_RDWR);
        write(fd, a, 300);
        close(fd);

        fd = open(FILE,O_RDWR);
        write(fd, buf, 100);
        close(fd);

This still produces a file with 100 bytes, rather than the expected size of 300 bytes as created by the first write.

**Comment by CamStan** [comment]
Created a PR to add a test and reproduce this issue (#510).

**Comment by clmendes** [comment]
@MichaelBrim , @adammoody : I tested the new-margotree branch today, by building Unify with it on Quartz under ~mendes3/UnifyFS-Mike/ , and indeed it seems to fix the problem in this issue: the resulting exported file has 300 bytes!

I will wait to close this issue until the fix is formally merged into Unify's dev.

**Comment by jenest** [comment]
This appears to be fixed. Closing issue. Thanks!
